File: examples/typescript/canonicalizeHashSign.ts
/**
 * Ledger canonical JSON + SHA-256 chain hash + ed25519 sign/verify.
 *
 * TypeScript / Node reference for docs/LEDGER_API.md.
 * Matches Python 3.10 json.dumps(separators=(',', ':'), sort_keys=True, ensure_ascii=False).
 *
 *   npx tsx examples/typescript/canonicalizeHashSign.ts
 *
 * Signing uses Node Web Crypto Ed25519 (Node 19+) by default.
 * Optional isomorphic lib: npm install @noble/ed25519 (or tweetnacl) — same raw 32-byte keys.
 *
 * TEST-ONLY: the demo runner generates an ephemeral keypair, prints the public key,
 * signs vector-1, verifies, and drops the private key. Nothing is written to disk.
 */

import { createHash, webcrypto } from "node:crypto";
import fs from "node:fs";
import path from "node:path";

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

function sortKeys(value: JsonValue): JsonValue {
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (value !== null && typeof value === "object") {
    const out: { [key: string]: JsonValue } = {};
    for (const key of Object.keys(value).sort()) {
      out[key] = sortKeys(value[key]!);
    }
    return out;
  }
  return value;
}

/** Deterministic JSON string: sorted keys, separators (",", ":"), UTF-8 / ensure_ascii=false. */
export function canonicalJson(obj: JsonValue): string {
  return JSON.stringify(sortKeys(obj));
}

/**
 * SHA-256 lowercase hex of:
 *   prev_hash_bytes + canonical_entity_bytes + timestamp_bytes + actor_id_bytes
 * null / "" prev_hash → empty prev_hash_bytes.
 */
export function computeHash(
  prevHash: string | null | undefined,
  entity: JsonValue,
  timestamp: string,
  actorId: string,
): string {
  const prev = prevHash ? Buffer.from(prevHash, "utf8") : Buffer.alloc(0);
  const data = Buffer.concat([
    prev,
    Buffer.from(canonicalJson(entity), "utf8"),
    Buffer.from(timestamp, "utf8"),
    Buffer.from(actorId, "utf8"),
  ]);
  return createHash("sha256").update(data).digest("hex");
}

export async function signHashEd25519(
  privateKey: CryptoKey,
  currentHashHex: string,
): Promise<string> {
  const sig = await webcrypto.subtle.sign(
    "Ed25519",
    privateKey,
    Buffer.from(currentHashHex, "utf8"),
  );
  return Buffer.from(sig).toString("base64");
}

export async function verifySignature(
  publicKeyBytes: Uint8Array,
  signatureB64: string,
  currentHashHex: string,
): Promise<boolean> {
  try {
    const key = await webcrypto.subtle.importKey(
      "raw",
      publicKeyBytes,
      "Ed25519",
      false,
      ["verify"],
    );
    return webcrypto.subtle.verify(
      "Ed25519",
      key,
      Buffer.from(signatureB64, "base64"),
      Buffer.from(currentHashHex, "utf8"),
    );
  } catch {
    return false;
  }
}

export async function generateTestKeypair(): Promise<{
  privateKey: CryptoKey;
  publicKeyBytes: Uint8Array;
}> {
  const pair = await webcrypto.subtle.generateKey("Ed25519", true, [
    "sign",
    "verify",
  ]);
  const publicKeyBytes = new Uint8Array(
    await webcrypto.subtle.exportKey("raw", pair.publicKey),
  );
  return { privateKey: pair.privateKey, publicKeyBytes };
}

export type LedgerVector = {
  id: string;
  entity: JsonValue;
  prev_hash: string | null;
  timestamp?: string;
  ledger_timestamp?: string;
  actor_id: string;
  canonical_json: string;
  expected_hash_hex?: string;
  expected_sha256_hex?: string;
};

export type VectorDoc = {
  vectors: LedgerVector[];
  verify_fixture_test_only?: {
    vector_id: string;
    public_key_base64: string;
    signature_base64: string;
  };
};

export function vectorTimestamp(v: LedgerVector): string {
  return v.timestamp || v.ledger_timestamp || "";
}

export function vectorExpectedHash(v: LedgerVector): string {
  return v.expected_hash_hex || v.expected_sha256_hex || "";
}

function loadVectors(root: string): VectorDoc {
  const vectorsPath = path.join(root, "tests", "ledger_vectors", "test_vectors.json");
  if (!fs.existsSync(vectorsPath)) {
    throw new Error(`test_vectors.json not found at ${vectorsPath}`);
  }
  return JSON.parse(fs.readFileSync(vectorsPath, "utf8")) as VectorDoc;
}

export async function checkVectors(doc: VectorDoc): Promise<number> {
  let failed = 0;
  for (const v of doc.vectors) {
    const gotCanon = canonicalJson(v.entity);
    const expected = vectorExpectedHash(v);
    const gotHash = computeHash(
      v.prev_hash,
      v.entity,
      vectorTimestamp(v),
      v.actor_id,
    );
    console.log(`${v.id}: canonical=${gotCanon}`);
    console.log(`${v.id}: sha256=${gotHash}`);
    if (gotCanon !== v.canonical_json) {
      console.error(`${v.id}: FAIL canonical JSON mismatch`);
      failed += 1;
    }
    if (gotHash !== expected) {
      console.error(`${v.id}: FAIL hash mismatch expected ${expected}`);
      failed += 1;
    }
  }
  return failed;
}

function repoRoot(): string {
  const argv1 = process.argv[1]?.replace(/\\/g, "/") ?? "";
  if (/(^|\/)canonicalizeHashSign\.(ts|js)$/.test(argv1)) {
    return path.resolve(path.dirname(process.argv[1]!), "..", "..");
  }
  return process.cwd();
}

async function main(): Promise<void> {
  const doc = loadVectors(repoRoot());
  let failed = await checkVectors(doc);

  const fixture = doc.verify_fixture_test_only;
  if (fixture) {
    const vector = doc.vectors.find((v) => v.id === fixture.vector_id);
    if (!vector) {
      console.error("FAIL verify fixture vector missing");
      failed += 1;
    } else {
      const pk = Buffer.from(fixture.public_key_base64, "base64");
      const ok = await verifySignature(
        pk,
        fixture.signature_base64,
        vectorExpectedHash(vector),
      );
      console.log(`verify_fixture_test_only (${fixture.vector_id}): ${ok}`);
      if (!ok) failed += 1;
    }
  }

  const { privateKey, publicKeyBytes } = await generateTestKeypair();
  console.log("\nGenerated ephemeral ed25519 keypair (TEST-ONLY; not saved)");
  console.log("public_key_base64:", Buffer.from(publicKeyBytes).toString("base64"));
  const first = doc.vectors[0]!;
  const h = computeHash(
    first.prev_hash,
    first.entity,
    vectorTimestamp(first),
    first.actor_id,
  );
  const sig = await signHashEd25519(privateKey, h);
  console.log("ephemeral_signature_base64 (vector-1):", sig);
  console.log(
    "ephemeral_verify_ok:",
    await verifySignature(publicKeyBytes, sig, h),
  );

  if (failed) {
    process.exit(1);
  }
}

const argv1 = process.argv[1]?.replace(/\\/g, "/") ?? "";
if (/(^|\/)canonicalizeHashSign\.(ts|js)$/.test(argv1)) {
  main().catch((err: unknown) => {
    console.error(err);
    process.exit(1);
  });
}

File: examples/python/canonicalize_hash_sign.py
#!/usr/bin/env python3
"""
Reference implementation: canonical JSON -> SHA256 chain hash -> ed25519 sign/verify (demo/test only)

Usage:
  pip install pynacl
  python3 examples/python/canonicalize_hash_sign.py

This script provides:
- canonical_json_bytes(obj)
- compute_current_hash(prev_hash, canonical_entity_bytes, timestamp, actor_id)
- sign_current_hash(private_key_bytes, current_hash_hex) -> base64 signature
- verify_signature(public_key_bytes, signature_b64, current_hash_hex) -> bool

Note: In production use a KMS/HSM for signing. This script generates an ephemeral test keypair if libs are available.
"""

import json
import hashlib
import base64
from typing import Any

try:
    from nacl.signing import SigningKey, VerifyKey
    from nacl.exceptions import BadSignatureError
    HAVE_NACL = True
except Exception:
    HAVE_NACL = False


def canonical_json_bytes(obj: Any) -> bytes:
    """Deterministic canonical JSON bytes as defined in LEDGER_API.md"""
    return json.dumps(obj, separators=(',', ':'), sort_keys=True, ensure_ascii=False).encode('utf-8')


def compute_current_hash(prev_hash: str, canonical_entity_bytes: bytes, timestamp: str, actor_id: str) -> str:
    """Compute SHA256 hex of prev_hash_bytes + canonical_entity_bytes + timestamp_bytes + actor_id_bytes"""
    prev = (prev_hash or '').encode('utf-8')
    data = prev + canonical_entity_bytes + timestamp.encode('utf-8') + actor_id.encode('utf-8')
    return hashlib.sha256(data).hexdigest()


def sign_current_hash_ed25519(private_key_bytes: bytes, current_hash_hex: str) -> str:
    """Sign the ASCII hex bytes of current_hash using ed25519 and return base64 signature"""
    if not HAVE_NACL:
        raise RuntimeError('PyNaCl required for signing (pip install pynacl)')
    sk = SigningKey(private_key_bytes)
    sig = sk.sign(current_hash_hex.encode('utf-8')).signature
    return base64.b64encode(sig).decode('ascii')


def verify_signature_ed25519(public_key_bytes: bytes, signature_b64: str, current_hash_hex: str) -> bool:
    if not HAVE_NACL:
        raise RuntimeError('PyNaCl required for verification (pip install pynacl)')
    vk = VerifyKey(public_key_bytes)
    sig = base64.b64decode(signature_b64)
    try:
        vk.verify(current_hash_hex.encode('utf-8'), sig)
        return True
    except Exception:
        return False


def generate_test_keypair():
    if not HAVE_NACL:
        raise RuntimeError('PyNaCl required to generate test keypair (pip install pynacl)')
    sk = SigningKey.generate()
    pk = sk.verify_key
    return sk.encode(), pk.encode()


if __name__ == '__main__':
    import pathlib
    root = pathlib.Path(__file__).resolve().parents[2]
    vectors_path = root / 'tests' / 'ledger_vectors' / 'test_vectors.json'
    try:
        with open(vectors_path, 'r', encoding='utf-8') as f:
            vectors = json.load(f)
    except FileNotFoundError:
        print('No test_vectors.json found at', vectors_path)
        vectors = []

    print('Computing hashes for', len(vectors), 'vectors')
    for i, v in enumerate(vectors, start=1):
        canonical = canonical_json_bytes(v['entity'])
        h = compute_current_hash(v.get('prev_hash', ''), canonical, v['ledger_timestamp'], v['actor_id'])
        print(f"VECTOR {i}: id={v['entity'].get('id')} hash={h}")

    if HAVE_NACL and vectors:
        sk_bytes, pk_bytes = generate_test_keypair()
        print('\nGenerated test ed25519 keypair (TEST-ONLY)')
        print('public key (base64):', base64.b64encode(pk_bytes).decode('ascii'))
        # demonstrate signing first vector
        first = vectors[0]
        first_hash = compute_current_hash(first.get('prev_hash', ''), canonical_json_bytes(first['entity']), first['ledger_timestamp'], first['actor_id'])
        sig = sign_current_hash_ed25519(sk_bytes, first_hash)
        print('example signature (base64) for first vector:', sig)
        ok = verify_signature_ed25519(pk_bytes, sig, first_hash)
        print('verification ok?', ok)
